src/mscorlib/src/System/TimeZoneInfo.cs

```diff
@@ -1944,10 +1971,173 @@ private class OffsetAndRule {
+            {
+                rawData = null;
+                id = null;
+                return false;
+            }
+
+            try
+            {
+                rawData = File.ReadAllBytes(tzFilePath);
+                if (rawData != null && string.IsNullOrEmpty(id))
+                {
+                    id = FindTimeZoneId(rawData);
+                }
+
+                return rawData != null;
+            }
+            catch (IOException)
+            {
+                rawData = null;
+                id = null;
+                return false;
+            }
+        }
+
+        private static string GetTzEnvironmentVariable()
+        {
+            string result = Environment.GetEnvironmentVariable("TZ");
+            if (!string.IsNullOrEmpty(result))
+            {
+                if (result[0] == ':')
+                {
+                    // strip off the ':' prefix
+                    result = result.Substring(1);
+                }
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Find the time zone id by searching all the tzfiles for the one that matches rawData
+        /// and return its file name.
+        /// </summary>
+        private static string FindTimeZoneId(byte[] rawData)
+        {
+            string id = null;
+            string localtimeFilePath = Path.Combine(c_timeZoneDirectory, "localtime");
+
+            foreach (string filePath in Directory.EnumerateFiles(c_timeZoneDirectory, "*", SearchOption.AllDirectories))
+            {
+                // skip the localtime file, since that won't give us the correct id
+                if (filePath != localtimeFilePath)
+                {
+                    try
+                    {
+                        using (FileStream stream = new FileStream(filePath, FileMode.Open, FileAccess.Read))
+                        {
+                            int fileLength = (int)stream.Length;
+                            if (fileLength == rawData.Length)
+                            {
+                                int index = 0;
+                                int count = fileLength;
+                                byte[] bytes = new byte[count];
+                                while (count > 0)
+                                {
+                                    int n = stream.Read(bytes, index, count);
+                                    if (n == 0)
+                                        __Error.EndOfFile();
+                                    index += n;
+                                    count -= n;
+                                }
+
+                                for (int i = 0; i<fileLength; i++)
+                                {
+                                    if (bytes[i] != rawData[i])
+                                    {
+                                        continue;
+                                    }
+                                }
```

Comment:
Rather than doing a second pass through, could we just do the comparison as we go as part of reading through the file?  I don't know how big these files typically are, but if they're relatively large, that would also permit the use of a smaller buffer.  And because the buffer wouldn't need to be the same size as the file, the allocation of the buffer could be moved to outside of all of these loops so that the same buffer is used across all files rather than allocated for each.